API: Make 64bit default integer on 64bit windows

seberg opened the pull request:
**This is a draft to get the ball rolling.  @Kai-Striega was hoping to probe this a bit more in downstream packages, but anyone excited is very much invited to join in and test downstream or ~start fixing tests (all green now)~ look at the changes, especially the random ones)**

### User Facing changes:

This changes the our default integer to `intp`, which really only changes behavior on 64bit windows.  This also means:
* `np.int_` and `np.uint` are now aliases to `intp` effectively (maybe we want to simplify this in some follow-up.
* `np.long` and `np.ulong` now exist
* (I am accepting intp is not ideal and may itself be better as `Py_ssize_t`, it isn't relevant except on a single very niche platform.  Also, I kinda want to just change it in C...)

The legacy *random* API is _untouched_ here (i.e. would continue using the old `long`), many many tests will fail due to this on windows and there are probably still holes.

Of course, and some ufunc loops need adapting (use int64), this isn't user-facing though.

### API changes for C-API/Cython users

Cython users may get a small problem, but they are also the ones complaining the most :).  We have to remove `np.int_t` because it cannot be defined for both 1.x and 2.x at the same time.
Both in Cython and in C there is a new `NPY_DEFAULT_INT` type-code (although I suspect cython users have easier paths).  They may have to use a fused-type (themselves) or simply add a manual casts.

### Other changes:

* `NPY_INTP` now prefers mapping to long rather than int (if same).  I think this makes a lot more sense (we always prefer long), but is a small API change/incompatibility (runtime type number of intp may not match the compile time one on NumPy 1.x when compiling with/for NumPy 2.x). *Mainly: Should be mentioned in release note*
* There is a new constant added next to the API table to fetch the NumPy C-version from our macros.  I think this is fine, NumPy 1.x would require a function call otherwise.

seberg committed:
API: Make intp the default integer in NumPy core

seberg committed:
API: Remove `int_` and `uint_` from Cython as they are misleading as long

seberg committed:
MAINT: Keep using `long` in for legacy random number generation

The new random number generation defaults to int64 throughout.

seberg committed:
fixup cython

seberg committed:
fixup initial change

seberg committed:
API: Add NPY_DEFAULT_INT which is runtime (unless NumPy >2 is forced)

Also test it within cython

seberg committed:
fixup initial (add/multiply reduction)

seberg committed:
Fixup NPY_DEFAULT_INT definition (and version exposure)

seberg committed:
ENH: Allow int64 explicitly in `ldexp`

This way, any integer input should work fine, in particular any 64
bit platform.  Chosing this rather than intp, since there seems
little reason to not support 64bit ints on 32bit platforms.

seberg committed:
TST: Adapt _ufunc_tests to use intp in ufuncs (default integer)

seberg committed:
API: rewire/rename np.int_ and np.uint scalars and add np.long/np.ulong

Some tests will be failing and docs will need update probably.

seberg committed:
BUG: Fix (probably) legacy binomial and use `np.long` in random

seberg committed:
TST: Adapt some tests to ensure long use and test default int a bit

seberg committed:
TST: Fix ufunc type resolution test for long (and a bit more)

seberg commented:
Fixes welcomem ;).  Let's see, many are just direct adaptations to the change.  Lot's of failures in `random.binomial`, but hopefully that is fixed now (certainly will be more in random).

seberg committed:
TST: Fix legacy random test hashes to toggle on long (not default int)

seberg committed:
TST: Remove duplicated test that needed adaption for default int change

seberg committed:
TST: Fix inplace_add ufunc test by using intp/default int

Otherwise, fixing the rest of the test seems actually more annoying.

seberg committed:
TST: Adjust random test for long (not default int) return and branching

seberg committed:
Use intp in np.random.binomial for inputs (core uses int64 always)

This is just minimal, really should use int64, but this is the legacy
API anyway.

seberg committed:
BUG: Fix previously unused int64/uint64 in umath generation for cross compile

(emscriptn CI was failing due to this)

seberg committed:
BUG: Another random path that should probably keep using long

(Might be good to have more tests, seemed mainly tested indirectly
via `choice` tests.)

seberg committed:
TST: One more small test fixup for changed default dtype

seberg commented:
Except legacy `np.random`, things should be settling.  Some of the random legacy are probably fixed a bit incorrectly right now, the random changes are really the trickest part here!

Not changing the legacy results, means that some internal int64s are cast to long. To allow default integer input, we cannot safe-cast user input to `long` anymore.  We will need to cast it to `int64` and probably but check that the values fit into `long`.
(Otherwise, we force users to ensure inputs are long, which seems like too much potential churn.)

seberg committed:
BUG,API: like int32, intp must map to long if long and int are same size

Otherwise things get confusing, the default integer must be long after
all if long and int are the same size.

seberg committed:
MAINT: Fix hypergeometric to use INT64, but raise if values surpass long

Also does this for bernoulli, its a mixed bag, many legacy ones use
the `disc()` which just force-casts from int64 to long anyway...

seberg committed:
Try with py_ssize_t (I am very unclear why float doesn't fail earlier)

seberg committed:
Simply us Python `int()`...

Kai-Striega commented:
I spent the last couple of hours trying to build SciPy from source on Windows by following the [CI](https://github.com/scipy/scipy/blob/main/.github/workflows/windows.yml). Not being a Windows person - it's proving to be quite the challenge. I think this is due to my lack of experience on Windows, not due to your changes. Tomorrow I'll probably take another look and see if I can make some progress

seberg commented:
@bashtage can I snipe you to have a look at the (admittedly not polished) changes in random here?  Dealing with the legacy random code seems like one of the more hairy points potentially.  Maybe it's all fine (just a bit rough), but I am not sure about how we would want to do it here.

bashtage commented on numpy/random/mtrand.pyx:
```diff
@@ -3481,16 +3506,20 @@ cdef class RandomState:
         _dp = PyFloat_AsDouble(p)
-        _in = <long>n
+        _in = int(n)
         check_constraint(_dp, 'p', CONS_BOUNDED_0_1)
```
Is the a python function call rather than a C cast? Maybe Cython 3 does the right thing here?

seberg replied:
The problem was that this can be a Python float, somehow the `<long>` cast uses a long cast that allows floats, while other to integer casts don't.  This _does_ depend on the Cython version (it worked locally).

I suspect it's the opposite, it was necessary in Cython 3 to do something more than a cast.

Will try once more locally, but maybe the best path is to just extract the value from the array (which we have created here in either case).

bashtage replied:
Could you just case twice? `<long>(<int>n)` to avoid python?

bashtage commented on numpy/random/mtrand.pyx:
```diff
@@ -3933,15 +3968,18 @@ cdef class RandomState:
-        ongood = <np.ndarray>np.PyArray_FROM_OTF(ngood, np.NPY_LONG, np.NPY_ALIGNED)
-        onbad = <np.ndarray>np.PyArray_FROM_OTF(nbad, np.NPY_LONG, np.NPY_ALIGNED)
-        onsample = <np.ndarray>np.PyArray_FROM_OTF(nsample, np.NPY_LONG, np.NPY_ALIGNED)
+        ongood = <np.ndarray>np.PyArray_FROM_OTF(ngood, np.NPY_INT64, np.NPY_ALIGNED)
```
This seems a bit surprising in terms of the behavior on 32bit windows which I think should have a 32but interfer, if I understand the rule correctly.

seberg replied:
The internal machinery requires int64, and there was a cast to int64 below.  The thing that this _does_ relax, is that previous an int64 input would raise because it cannot be safely cast to long (int32) on windows and 32bits.

(I am happy to change it to intp though!  Just seemed like might as well go to int64 directly.)

rkern replied:
But that _was_ the point of using `NPY_LONG` here, according to the comment above it. To error out early rather than late.

seberg replied:
@rkern right, I should remove the comment.  Ensuring an error is why I added the explicit check below.  This ensures that:
1. We still get an error when the value is out of bounds.  (later, but not much)
2. The user can pass a default-integer array.  This would otherwise simply fail otherwise because it is not a safe cast.

So, the relaxation and explicit check seemed unfortunately necessary because I don't want to break default integer array input.  Or am I misunderstanding your comment?

pllim commented:
Hello! We only test nightly numpy on Linux. Is this ready for downstream testing? Actually, this PR has no wheel, so I am not sure how easy it is to build this from scratch and test locally on Windows. But `astropy` would be interested to test this at some point.

seberg commented:
> Is this ready for downstream testing?

Yes, I think the NumPy changes should be mostly settled so that would make me very happy!

Just triggered the build, are the artifacts visible to you: https://github.com/numpy/numpy/actions/runs/5645884112#artifacts? just in case it works better [windows-py10](https://github.com/numpy/numpy/suites/14521621389/artifacts/821394315) and [windows-py11](https://github.com/numpy/numpy/suites/14521621389/artifacts/821394321).

I do not know how badly this will affect astropy or your dependencies; I would think some cython code breaks because it uses `npc.int_[::1]` or `long[::1]` memoryviews expecting/knowing the default integer input should come in.  (Which is of course the exact opposite that many small libraries fight with, who hard-code int64, presuming that everyone uses that, but win64 users are left unable to use their libs).

pllim commented:
Thanks!

seberg committed:
BUG: Ensure `randint` returns a Python integer by default

This was relied upon by sklearn and it seems reasonable to still
default to returning a Python integer.

seberg commented:
@mtsokol, I had added `long` and `ulong` here, because we effectively used `int_` for that before and that is a worse name.  I can live with requiring `np.dtype("long").type` also, mainly there is quite a bit of merge conflict in the parts related to that addition and I was wondering if things changed a bit w.r.t. to adding them.

(Maybe rebasing should wait until the type alias refactor is done...)

mtsokol committed:
Adjust type aliases after rebase

mtsokol committed:
Apply linting

seberg committed:
TST: Fix random test for `dtype=int` being special cased in mtrand

seberg commented:
Thanks for updating @mtsokol!  There are two issues remaining here, if I am not missing something big:
1. It would be nice to explicitly vet the choices in `mtrand.pyx`.  We can't change the default, but how exactly to do it with `dtype=int`, etc. is not clear.
2. We need to add typing for the new long (and fix a few typing tests).  Maybe @BvB93 can have a look.  We may look into splitting out adding `np.long` here for simplicity though.  (Adding `np.long` itself is simple, but I am not sure how involved the typing is.)

BvB93 commented:
> We need to add typing for the new long (and fix a few typing tests). Maybe @BvB93 can have a look. We may look into splitting out adding np.long here for simplicity though. (Adding np.long itself is simple, but I am not sure how involved the typing is.)

The typing sides of things is fortunately not too difficult: `(u)int` just becomes a `(u)intp` alias, and `(u)long` reuses the old `(u)int` type alias (+ a handful a `np.ctypeslib`-related test fixes). Got a small patch down below, though I can also directly push to your branch of you prefer.

<details>
  <summary>typing patch</summary>

```diff
diff --git a/numpy/__init__.pyi b/numpy/__init__.pyi
index 418bdf614..03b118fdf 100644
--- a/numpy/__init__.pyi
+++ b/numpy/__init__.pyi
@@ -2857,7 +2857,8 @@ def __init__(self, value: _IntValue = ..., /) -> None: ...
 short = signedinteger[_NBitShort]
 intc = signedinteger[_NBitIntC]
 intp = signedinteger[_NBitIntP]
-int_ = signedinteger[_NBitInt]
+int_ = intp
+long = signedinteger[_NBitInt]
 longlong = signedinteger[_NBitLongLong]
 
 # TODO: `item`/`tolist` returns either `dt.timedelta` or `int`
@@ -2938,7 +2939,8 @@ def __init__(self, value: _IntValue = ..., /) -> None: ...
 ushort = unsignedinteger[_NBitShort]
 uintc = unsignedinteger[_NBitIntC]
 uintp = unsignedinteger[_NBitIntP]
-uint = unsignedinteger[_NBitInt]
+uint = uintp
+ulong = unsignedinteger[_NBitInt]
 ulonglong = unsignedinteger[_NBitLongLong]
 
 class inexact(number[_NBit1]):  # type: ignore
diff --git a/numpy/typing/mypy_plugin.py b/numpy/typing/mypy_plugin.py
index f4ad55341..78fea240f 100644
--- a/numpy/typing/mypy_plugin.py
+++ b/numpy/typing/mypy_plugin.py
@@ -59,7 +59,7 @@ def _get_precision_dict() -> dict[str, str]:
         ("_NBitShort", np.short),
         ("_NBitIntC", np.intc),
         ("_NBitIntP", np.intp),
-        ("_NBitInt", np.int_),
+        ("_NBitInt", np.long),
         ("_NBitLongLong", np.longlong),
 
         ("_NBitHalf", np.half),
diff --git a/numpy/typing/tests/data/reveal/ctypeslib.pyi b/numpy/typing/tests/data/reveal/ctypeslib.pyi
index a9712c074..5c3b2138f 100644
--- a/numpy/typing/tests/data/reveal/ctypeslib.pyi
+++ b/numpy/typing/tests/data/reveal/ctypeslib.pyi
@@ -79,17 +79,9 @@
 assert_type(np.ctypeslib.as_array(1), npt.NDArray[Any])
 assert_type(np.ctypeslib.as_array(pointer), npt.NDArray[Any])
 
-if sys.platform == "win32":
-    assert_type(np.ctypeslib.as_ctypes_type(np.int_), type[ct.c_int])
-    assert_type(np.ctypeslib.as_ctypes_type(np.uint), type[ct.c_uint])
-    assert_type(np.ctypeslib.as_ctypes(AR_uint), ct.Array[ct.c_uint])
-    assert_type(np.ctypeslib.as_ctypes(AR_int), ct.Array[ct.c_int])
-    assert_type(np.ctypeslib.as_ctypes(AR_uint.take(0)), ct.c_uint)
-    assert_type(np.ctypeslib.as_ctypes(AR_int.take(0)), ct.c_int)
-else:
-    assert_type(np.ctypeslib.as_ctypes_type(np.int_), type[ct.c_long])
-    assert_type(np.ctypeslib.as_ctypes_type(np.uint), type[ct.c_ulong])
-    assert_type(np.ctypeslib.as_ctypes(AR_uint), ct.Array[ct.c_ulong])
-    assert_type(np.ctypeslib.as_ctypes(AR_int), ct.Array[ct.c_long])
-    assert_type(np.ctypeslib.as_ctypes(AR_uint.take(0)), ct.c_ulong)
-    assert_type(np.ctypeslib.as_ctypes(AR_int.take(0)), ct.c_long)
+assert_type(np.ctypeslib.as_ctypes_type(np.int_), type[ct.c_long])
+assert_type(np.ctypeslib.as_ctypes_type(np.uint), type[ct.c_ulong])
+assert_type(np.ctypeslib.as_ctypes(AR_uint), ct.Array[ct.c_ulong])
+assert_type(np.ctypeslib.as_ctypes(AR_int), ct.Array[ct.c_long])
+assert_type(np.ctypeslib.as_ctypes(AR_uint.take(0)), ct.c_ulong)
+assert_type(np.ctypeslib.as_ctypes(AR_int.take(0)), ct.c_long)

```
</details>

seberg committed:
DOC: Rephrase now incorrect comment in mtrand.pyx

seberg committed:
TYP: Fixup typing for long changes

seberg committed:
TST,BUG: Work around random bug in tests

seberg committed:
TST: Fix test due to changes on main

seberg committed:
TST: Fix randint test the right way

The test should check for the result type, but the way I had adapted
it was thus the wrong way around (and didn't work)

seberg commented:
@mtsokol was looking into some typing updates, as chances are that a few typing stubs will be incorrect in the current state.  I would propose to follow-up on these.

More importantly, it would be good to move forward here since downstream will be affected, even if mildly hopefully, so it would be good to move forward soon.
So, as I asked many times in the meetings:  @ngoldbaum, @charris, and everyone else, please have a look, so that we can move forward with this soon.

bashtage commented on numpy/random/mtrand.pyx:
```diff
@@ -3471,7 +3493,11 @@ cdef class RandomState:
                     _dp = (<double*>np.PyArray_MultiIter_DATA(it, 1))[0]
-                    _in = (<long*>np.PyArray_MultiIter_DATA(it, 2))[0]
+                    _in = (<np.npy_intp*>np.PyArray_MultiIter_DATA(it, 2))[0]
+                    if _in > LONG_MAX:
```
The check you added here could be implemented using the check_array_constraint above by creating a new constant such as `CONS_NON_NEGATVE_INBOUNDS_LONG`.

seberg replied:
OK, added a commit for this as `LEGACY_CONS_NON_NEGATVE_INBOUNDS_LONG`.  It is slightly awkward, because the scalar check is typed as `double`.
That is fine in practice since we are only interested in longs that are 32bit (and thus fit double).
(on both sides, I used the knowledge that this only makes sense for integers.)

bashtage commented on numpy/random/mtrand.pyx:
```diff
@@ -3954,10 +3993,11 @@ cdef class RandomState:
+            raise ValueError(
+                    "`nsample` is out of bounds for long, consider using "
+                    "the new generator API for 64bit integers.")
+
```
There is a constraint check for `onsample` below in `discrete_broadcast_iii`. Could avoid an extra check by rolling that into the existing check.

bashtage reviewed (commented):
Some small things about constraint checking.

seberg committed:
MAINT: Move Long-bound checks into helper

This is somewhat strange, because we cast to double in the scalar
case.  This is OK in practice because we are really only interested
in 32bit longs, which are guaranteed to fit into double...

ngoldbaum commented on numpy/__init__.pxd:
```diff
@@ -699,12 +700,7 @@ ctypedef npy_float64    float64_t
-ctypedef npy_long       int_t
 ctypedef npy_longlong   longlong_t
-
-ctypedef npy_ulong      uint_t
```
I guess we don't need a cython type name for `np.long` because they can just use `long`?

ngoldbaum commented on numpy/_core/code_generators/generate_umath.py:
```diff
@@ -94,7 +94,8 @@ def finish_signature(self, nin, nout):
 
 def _check_order(types1, types2):
-    dtype_order = allP + "O"
+    # Insert kK (int64) after all other ints (assumes long long isn't larger)
```
Can you add a sentence explaining why int64 needs to be last? It's not clear to me just reading this over.

seberg replied:
kK is assumed to be the largest int.  This function checks the order.  Let's put a sentence of doc to the function, I suspect that should solve this also.

ngoldbaum commented on numpy/_core/code_generators/generate_umath.py:
```diff
@@ -275,6 +276,10 @@ def english_upper(s):
+    # We sometimes need int64, but we have no obvious char for it, use k and
+    # define it as `int64` below.
+    'k': 'int64',
+    'K': 'uint64',
```
It's not clear to me how [this page](https://numpy.org/doc/1.26/reference/arrays.scalars.html) get generated - but maybe it needs to be rearranged now that `int64` isn't a sized alias anymore? And will 'k' show up somewhere on that page?

seberg replied:
`k` isn't a thing, it is only used in this file internally to define some ufunc types.  That said, you are right that I didn't really change the docs yet, and that page looks like it needs to be expanded to list the new `long`!

ngoldbaum replied:
Ah, it wasn't clear to me it wasn't user-facing. Thanks!

ngoldbaum commented on numpy/random/mtrand.pyx:
```diff
@@ -622,8 +623,13 @@ cdef class RandomState:
+        [0, ``np.iinfo("long").max``].
+
+        .. warning::
+           This function users the C-long dtype, which is 32bit on windows
```
```suggestion
           This function uses the C-long dtype, which is 32bit on windows
```

ngoldbaum reviewed (commented):
I can't speak to the random changes but I spotted some minor things, see below.

It would be nice to have some discussion about this in the NumPy 2.0 migration guide and needs a release note too.

seberg committed:
Update numpy/random/mtrand.pyx

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

seberg committed:
DOC: Add release note and transition guide entry

seberg commented:
Added the releasenote/transition docs.  `doc/source/reference/arrays.scalars.rst` needs updating, but I skipped that for  now because gh-25030 will just create conflicts one way or another.

mtsokol commented on doc/release/upcoming_changes/24224.change.rst:
```diff
@@ -0,0 +1,7 @@
+Default integer now 64bit on 64bit windows
+------------------------------------------
+The default NumPy integer is no 64bit on all 64bit systems as the
```
`is now 64bit`

mtsokol commented on numpy/typing/tests/data/reveal/ctypeslib.pyi:
```diff
@@ -79,17 +79,9 @@ assert_type(np.ctypeslib.as_array(AR_ubyte), npt.NDArray[np.ubyte])
+assert_type(np.ctypeslib.as_ctypes_type(np.ulong), type[ct.c_ulong])
+assert_type(np.ctypeslib.as_ctypes(AR_ulong), ct.Array[ct.c_ulong])
+assert_type(np.ctypeslib.as_ctypes(AR_long), ct.Array[ct.c_long])
+assert_type(np.ctypeslib.as_ctypes(AR_long.take(0)), ct.c_ulong)
```
It should be:
```
assert_type(np.ctypeslib.as_ctypes(AR_ulong.take(0)), ct.c_ulong)
```

mtsokol commented:
> @BvB93 do you know what is happening with the MyPy test, when I ran things locally it seemed to just skip all. Is the removal from the `np.int_` from the plugin somehow leading to this?

@seberg I added two comments - with these changes running `spin mypy` locally passes for me.

seberg committed:
Fix typing and typo based on review

seberg committed:
TYP: Try fixing/working around windows ctypeslib mypy failures

Simply sort the long codes before int, because it might mean we
prefer them (and that may fix the tests).  Also explicitly add
mapping to ssize_t for int (this pre-empts chaning the definition
of `intp` admittedly)

seberg commented:
Hmmm, that wasn't enough, going to try to swap the order in the ctypes definition and explicitly typing for `int`...

BvB93 commented:
Been trying to debug things locally for a bit, but running into issues wherein numpy is failing to build. `spin build -- -Dallow-noblas=true` runs without issue but the subsequent `spin mypy` raises when attempting to rebuild (?) numpy.

```
Invoking `build` prior to running tests:
$ 'C:\Users\hardd\miniconda3\envs\numpy\python.exe' vendored-meson/meson/meson.py compile -C build
Activating VS 17.6.5
INFO: automatically activated MSVC compiler environment
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: C:\Users\hardd\miniconda3\envs\numpy\Scripts\ninja.EXE -C C:/Users/hardd/Documents/GitHub/numpy/build
ninja: Entering directory `C:/Users/hardd/Documents/GitHub/numpy/build'
[1/1] Generating numpy/generate-version with a custom command
Saving version to numpy\version.py
$ 'C:\Users\hardd\miniconda3\envs\numpy\python.exe' vendored-meson/meson/meson.py install --only-changed -C build --destdir ../build-install
$ export PYTHONPATH="C:\Users\hardd\Documents\GitHub\numpy\build-install\usr\Lib\site-packages"
$ 'C:\Users\hardd\miniconda3\envs\numpy\python.exe' -P -c 'import numpy'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1354, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1316, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1256, in _find_spec
  File "C:\Users\hardd\miniconda3\envs\numpy\Lib\site-packages\_numpy_editable_loader.py", line 271, in find_spec
    tree = self.rebuild()
           ^^^^^^^^^^^^^^
  File "C:\Users\hardd\miniconda3\envs\numpy\Lib\site-packages\_numpy_editable_loader.py", line 312, in rebuild
    subprocess.run(self._build_cmd, cwd=self._build_path, env=env, stdout=stdout, check=True)
  File "C:\Users\hardd\miniconda3\envs\numpy\Lib\subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\hardd\miniconda3\envs\numpy\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\hardd\miniconda3\envs\numpy\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
NotADirectoryError: [WinError 267] The directory name is invalid
As a sanity check, we tried to import numpy.
Stopping. Please investigate the build error.
```

rgommers commented:
Hmm. The rebuild is expected, and that should take a fraction of a second if there are no changes to compiled code. This looks like a bug in `spin` to me, or maybe in `ninja` or in `meson-python`'s editable support. It's not clear from the error message what is going wrong - something is constructing an invalid path here, but that path isn't shown.

seberg commented:
Puh, typing works now... to a large degree just a full circle of changes in the end.  Typing changes here are now almost only a few small fix-ups.  I did a few greps over the typing stubs, but I cannot be 100% sure that there isn't a stray "default integer" around.  Although, it seems to me that typing things like array coercion is relatively limited, so use of `int` or `type[int]` is also.

So, I think this should be as ready as it gets, if the docs are fine.

ngoldbaum commented on doc/source/numpy_2_0_migration_guide.rst:
```diff
@@ -2,13 +2,48 @@
+Libraries interfacing with compiled code that are written in C, Cython, or
+a similar language may require updating to accomodate user input if they
+are using the ``long`` or equivalent type on the C-side.
+In this case, you may wish to using ``intp`` and cast user input or support
```
```suggestion
In this case, you may wish to use ``intp`` and cast user input or support
```

ngoldbaum commented on doc/source/numpy_2_0_migration_guide.rst:
```diff
@@ -2,13 +2,48 @@
+are using the ``long`` or equivalent type on the C-side.
+In this case, you may wish to using ``intp`` and cast user input or support
+both ``long`` and ``intp`` (to better support NumPy 1.x as well).
+When creating a new integer array in C or Cython, the new ``NPY_DEFAULT_INT``
```
Maybe we should define this in 1.26 too?

seberg replied:
Might make it a bit easier to use for downstream, but they still will need to force `numpy>=1.26` at compile time.  So I am tempted to say that unfortunately, it is better to vendor it:
```
#ifndef NPY_DEFAULT_INT
#define NPY_DEFAULT_INT NPY_LONG
#endif
```
(unless you force `>2.0` at compile time)?

Happy to follow-up with a backport, not sure it's helpful, but wouldn't hurt.

ngoldbaum reviewed (commented):
The docs look good, just a couple typo fixes and comments. @BvB93 can you give the typing changes one more look?

seberg committed:
Update doc/source/numpy_2_0_migration_guide.rst

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

BvB93 commented on numpy/typing/tests/data/reveal/ctypeslib.pyi:
```diff
@@ -80,16 +80,17 @@ assert_type(np.ctypeslib.as_array(1), npt.NDArray[Any])
+    assert_type(np.ctypeslib.as_ctypes(AR_ulong), ct.Array[ct.c_uint])
+    assert_type(np.ctypeslib.as_ctypes(AR_long), ct.Array[ct.c_int])
+    assert_type(np.ctypeslib.as_ctypes(AR_long.take(0)), ct.c_int)
+    assert_type(np.ctypeslib.as_ctypes(AR_ulong.take(0)), ct.c_uint)
 else:
-    assert_type(np.ctypeslib.as_ctypes_type(np.int_), type[ct.c_long])
-    assert_type(np.ctypeslib.as_ctypes_type(np.uint), type[ct.c_ulong])
-    assert_type(np.ctypeslib.as_ctypes(AR_uint), ct.Array[ct.c_ulong])
-    assert_type(np.ctypeslib.as_ctypes(AR_int), ct.Array[ct.c_long])
-    assert_type(np.ctypeslib.as_ctypes(AR_uint.take(0)), ct.c_ulong)
-    assert_type(np.ctypeslib.as_ctypes(AR_int.take(0)), ct.c_long)
+    assert_type(np.ctypeslib.as_ctypes_type(np.long), type[ct.c_long])
+    assert_type(np.ctypeslib.as_ctypes_type(np.ulong), type[ct.c_ulong])
+    assert_type(np.ctypeslib.as_ctypes(AR_ulong), ct.Array[ct.c_ulong])
+    assert_type(np.ctypeslib.as_ctypes(AR_long), ct.Array[ct.c_long])
+    assert_type(np.ctypeslib.as_ctypes(AR_long.take(0)), ct.c_long)
+    assert_type(np.ctypeslib.as_ctypes(AR_ulong.take(0)), ct.c_ulong)
```
LGTM here; nice to see you managed to get things working without commenting them out!

ngoldbaum commented:
Let's pull this in. Thanks for pushing this forward @seberg!

If you're looking at this PR because of new issues on Windows in CI or elsewhere, take a look at the [NumPy 2.0 Migration Guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html), which will be updated with specific information related to this change shortly after this PR is merged. In short, existing code that assumed the default integer type on windows is 32 bit may behave differently or raise new errors.

If that document doesn't answer your questions or if you are still confused, please feel free to [open an issue](https://github.com/numpy/numpy/issues/new?assignees=&labels=00+-+Bug&projects=&template=bug-report.yml&title=BUG%3A+%3CPlease+write+a+comprehensive+title+after+the+%27BUG%3A+%27+prefix%3E) describing your problem.
